src/charm.py

```diff
@@ -714,7 +766,9 @@ def _update_trusted_ca_certs(self, container: Container) -> bool:
             return False
 
         container.push(TRUSTED_CA_FILENAME, cert, make_dirs=True)
-        self._state.contract_cert_hash = cert_hash
+
+        if self.unit.is_leader():
+            self._state.contract_cert_hash = cert_hash
```

Comment:
Don't all the units need this cert in the workload container?
Ah on second thought, this part is just for storing the cert in state which only the leader can/should do.
But then higher up, if the cert is in the state, we won't push it to the workload container which seems like it could be an issue if a non-leader unit wants to talk to contracts.

I forget how the state works, is it per unit or per app? That will impact this I guess.

Reply:
You're right. This is possible. To help with this, we need to either:
1. Store the unit state.
2. Always refresh the cert.
3. Check if the cert if valid in the container, and refresh if it's not.

The second solution is the simplest, with a negligible burden of refreshing the certs and restarting the server. What do you think?

Reply:
Yeah I think option 2 is good.

Reply:
Just found out that it was actually working in this way all the time (I mean, always refreshing the cert). If you look closely:

```py
cert_hash = hash(cert)
if self._state.contract_cert_hash and cert_hash == self._state.contract_cert_hash:
    return False
```

This is just a `hash` function used by maps/dicts, not a stable hash algorithm (like SHA or MD5). So, it'll have a different value per Python runtime process. Since every hook runs in a separate Python runtime process, the condition in the if-statement is always `False`.

Since it won't break the actual behavior (which is luckily okay), I'll remove the state from here and always refresh the CA certs.

Reply:
Aha good catch!